File: app/src/main/java/com/kickstarter/ui/fragments/DiscoveryFragment.java
package com.kickstarter.ui.fragments;

import android.animation.AnimatorSet;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.jakewharton.rxbinding.view.RxView;
import com.kickstarter.R;
import com.kickstarter.libs.ActivityRequestCodes;
import com.kickstarter.libs.BaseFragment;
import com.kickstarter.libs.RecyclerViewPaginator;
import com.kickstarter.libs.RefTag;
import com.kickstarter.libs.SwipeRefresher;
import com.kickstarter.libs.qualifiers.RequiresFragmentViewModel;
import com.kickstarter.libs.utils.AnimationUtils;
import com.kickstarter.libs.utils.ViewUtils;
import com.kickstarter.models.Activity;
import com.kickstarter.models.Category;
import com.kickstarter.models.Project;
import com.kickstarter.services.DiscoveryParams;
import com.kickstarter.ui.ArgumentsKey;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.ActivityFeedActivity;
import com.kickstarter.ui.activities.LoginToutActivity;
import com.kickstarter.ui.activities.ProjectActivity;
import com.kickstarter.ui.activities.UpdateActivity;
import com.kickstarter.ui.adapters.DiscoveryAdapter;
import com.kickstarter.ui.data.LoginReason;
import com.kickstarter.viewmodels.DiscoveryFragmentViewModel;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.Bind;
import butterknife.ButterKnife;

import static com.kickstarter.libs.rx.transformers.Transformers.observeForUI;
import static com.kickstarter.libs.utils.TransitionUtils.slideInFromRight;
import static com.kickstarter.libs.utils.TransitionUtils.transition;

@RequiresFragmentViewModel(DiscoveryFragmentViewModel.ViewModel.class)
public final class DiscoveryFragment extends BaseFragment<DiscoveryFragmentViewModel.ViewModel> {
  private AnimatorSet heartsAnimation;
  private RecyclerViewPaginator recyclerViewPaginator;

  protected @Bind(R.id.discovery_empty_heart_filled) ImageView heartFilled;
  protected @Bind(R.id.discovery_empty_heart_outline) ImageView heartOutline;
  protected @Bind(R.id.discovery_empty_view) View emptyView;
  protected @Bind(R.id.discovery_hearts_container) View heartsContainer;
  protected @Bind(R.id.discovery_recycler_view) RecyclerView recyclerView;
  protected @Bind(R.id.disco_swipe_refresh_layout) SwipeRefreshLayout swipeRefreshLayout;

  public DiscoveryFragment() {}

  public static @NonNull DiscoveryFragment newInstance(final int position) {
    final DiscoveryFragment fragment = new DiscoveryFragment();
    final Bundle bundle = new Bundle();
    bundle.putInt(ArgumentsKey.DISCOVERY_SORT_POSITION, position);
    fragment.setArguments(bundle);
    return fragment;
  }

  @Override
  public @Nullable View onCreateView(final @NonNull LayoutInflater inflater, final @Nullable ViewGroup container,
    final @Nullable Bundle savedInstanceState) {
    super.onCreateView(inflater, container, savedInstanceState);
    final View view = inflater.inflate(R.layout.fragment_discovery, container, false);
    ButterKnife.bind(this, view);

    final DiscoveryAdapter adapter = new DiscoveryAdapter(this.viewModel.inputs);
    this.recyclerView.setAdapter(adapter);
    final LinearLayoutManager layoutManager = new LinearLayoutManager(this.recyclerView.getContext());
    this.recyclerView.setLayoutManager(layoutManager);
    new SwipeRefresher(
      this, this.swipeRefreshLayout, this.viewModel.inputs::refresh, this.viewModel.outputs::isFetchingProjects
    );
    this.recyclerViewPaginator = new RecyclerViewPaginator(this.recyclerView, this.viewModel.inputs::nextPage);

    this.viewModel.outputs.activity()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(adapter::takeActivity);

    this.viewModel.outputs.startHeartAnimation()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .filter(__ -> !lazyHeartCrossFadeAnimation().isRunning())
      .subscribe(__ -> lazyHeartCrossFadeAnimation().start());

    this.viewModel.outputs.projectList()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(adapter::takeProjects);

    this.viewModel.outputs.shouldShowEmptySavedView()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(show -> ViewUtils.setGone(this.emptyView, !show));

    this.viewModel.outputs.shouldShowOnboardingView()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(adapter::setShouldShowOnboardingView);

    this.viewModel.outputs.showActivityFeed()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(__ -> startActivityFeedActivity());

    this.viewModel.outputs.startUpdateActivity()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(this::startUpdateActivity);

    this.viewModel.outputs.startProjectActivity()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(projectAndRefTag -> this.startProjectActivity(projectAndRefTag.first, projectAndRefTag.second));

    this.viewModel.outputs.showLoginTout()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(__ -> this.startLoginToutActivity());

    RxView.clicks(this.heartsContainer)
      .compose(bindToLifecycle())
      .subscribe(__ -> this.viewModel.inputs.heartContainerClicked());

    return view;
  }

  @Override
  public void onPause() {
    super.onPause();
    this.heartsAnimation = null;
  }

  @Override
  public void onDetach() {
    super.onDetach();

    if (this.recyclerView != null) {
      this.recyclerView.setAdapter(null);
    }

    if (this.recyclerViewPaginator != null) {
      this.recyclerViewPaginator.stop();
    }
  }

  public boolean isAttached() {
    return this.viewModel != null;
  }

  public boolean isInstantiated() {
    return this.recyclerView != null;
  }

  private @NonNull AnimatorSet lazyHeartCrossFadeAnimation() {
    if (this.heartsAnimation == null) {
      this.heartsAnimation = AnimationUtils.INSTANCE.crossFadeAndReverse(this.heartOutline, this.heartFilled, 400L);
    }
    return this.heartsAnimation;
  }

  private void startActivityFeedActivity() {
    startActivity(new Intent(getActivity(), ActivityFeedActivity.class));
  }

  private void startLoginToutActivity() {
    final Intent intent = new Intent(getActivity(), LoginToutActivity.class)
      .putExtra(IntentKey.LOGIN_REASON, LoginReason.DEFAULT);
    startActivityForResult(intent, ActivityRequestCodes.LOGIN_FLOW);
    transition(getActivity(), slideInFromRight());
  }

  private void startProjectActivity(final @NonNull Project project, final @NonNull RefTag refTag) {
    final Intent intent = new Intent(getActivity(), ProjectActivity.class)
      .putExtra(IntentKey.PROJECT, project)
      .putExtra(IntentKey.REF_TAG, refTag);
    startActivity(intent);
    transition(getActivity(), slideInFromRight());
  }

  private void startUpdateActivity(final @NonNull Activity activity) {
    final Intent intent = new Intent(getActivity(), UpdateActivity.class)
      .putExtra(IntentKey.PROJECT, activity.project())
      .putExtra(IntentKey.UPDATE, activity.update());
    startActivity(intent);
    transition(getActivity(), slideInFromRight());
  }

  public void takeCategories(final @NonNull List<Category> categories) {
    this.viewModel.inputs.rootCategories(categories);
  }

  public void updateParams(final @NonNull DiscoveryParams params) {
    this.viewModel.inputs.paramsFromActivity(params);
  }

  public void clearPage() {
    this.viewModel.inputs.clearPage();
  }

  public void scrollToTop() {
    this.recyclerView.smoothScrollToPosition(0);
  }
}

File: app/src/main/java/com/kickstarter/ui/fragments/DiscoveryFragment.kt
package com.kickstarter.ui.fragments

import android.animation.AnimatorSet
import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.view.isGone
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.ConcatAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import com.kickstarter.R
import com.kickstarter.databinding.FragmentDiscoveryBinding
import com.kickstarter.libs.ActivityRequestCodes
import com.kickstarter.libs.RefTag
import com.kickstarter.libs.recyclerviewpagination.RecyclerViewPaginatorV2
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.AnimationUtils.crossFadeAndReverse
import com.kickstarter.libs.utils.ThirdPartyEventValues
import com.kickstarter.libs.utils.TransitionUtils
import com.kickstarter.libs.utils.ViewUtils
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.getEnvironment
import com.kickstarter.libs.utils.extensions.getPreLaunchProjectActivity
import com.kickstarter.libs.utils.extensions.getProjectIntent
import com.kickstarter.libs.utils.extensions.getSetPasswordActivity
import com.kickstarter.models.Activity
import com.kickstarter.models.Category
import com.kickstarter.models.Project
import com.kickstarter.services.DiscoveryParams
import com.kickstarter.ui.ArgumentsKey
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.activities.ActivityFeedActivity
import com.kickstarter.ui.activities.EditorialActivity
import com.kickstarter.ui.activities.LoginToutActivity
import com.kickstarter.ui.activities.UpdateActivity
import com.kickstarter.ui.adapters.DiscoveryActivitySampleAdapter
import com.kickstarter.ui.adapters.DiscoveryEditorialAdapter
import com.kickstarter.ui.adapters.DiscoveryOnboardingAdapter
import com.kickstarter.ui.adapters.DiscoveryProjectCardAdapter
import com.kickstarter.ui.data.Editorial
import com.kickstarter.ui.data.LoginReason
import com.kickstarter.ui.viewholders.EditorialViewHolder
import com.kickstarter.viewmodels.DiscoveryFragmentViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.launch

class DiscoveryFragment : Fragment() {
    private var heartsAnimation: AnimatorSet? = null
    private var recyclerViewPaginator: RecyclerViewPaginatorV2? = null

    private var binding: FragmentDiscoveryBinding? = null
    private var discoveryEditorialAdapter: DiscoveryEditorialAdapter? = null
    private val projectStarConfirmationString = R.string.project_star_confirmation
    private val disposables = CompositeDisposable()

    private lateinit var viewModelFactory: DiscoveryFragmentViewModel.Factory
    private val viewModel: DiscoveryFragmentViewModel.DiscoveryFragmentViewModel by viewModels { viewModelFactory }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        this.context?.getEnvironment()?.let { env ->
            viewModelFactory = DiscoveryFragmentViewModel.Factory(env)
        }
        binding = FragmentDiscoveryBinding.inflate(inflater, container, false)
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.RESUMED) {
                activity?.runOnUiThread { viewModel.inputs.fragmentLifeCycle(Lifecycle.State.RESUMED) }
            }
        }

        val discoveryActivitySampleAdapter = DiscoveryActivitySampleAdapter(this.viewModel.inputs)
        val discoveryEditorialAdapter = DiscoveryEditorialAdapter(this.viewModel.inputs)
        val discoveryOnboardingAdapter = DiscoveryOnboardingAdapter(this.viewModel.inputs)
        val discoveryProjectCardAdapter = DiscoveryProjectCardAdapter(this.viewModel.inputs)

        val discoveryAdapter = ConcatAdapter(discoveryOnboardingAdapter, discoveryEditorialAdapter, discoveryActivitySampleAdapter, discoveryProjectCardAdapter)
        this.discoveryEditorialAdapter = discoveryEditorialAdapter

        binding?.discoveryRecyclerView?.apply {
            adapter = discoveryAdapter
            layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
            recyclerViewPaginator = RecyclerViewPaginatorV2(
                this,
                { this@DiscoveryFragment.viewModel.inputs.nextPage() },
                this@DiscoveryFragment.viewModel.outputs.isFetchingProjects()
            )
        }

        binding?.discoverySwipeRefreshLayout?.let { swipeRefreshLayout ->
            swipeRefreshLayout.setOnRefreshListener {
                viewModel.inputs.refresh()
            }

            this.viewModel.outputs.isFetchingProjects()
                .compose(Transformers.observeForUIV2())
                .subscribe {
                    swipeRefreshLayout.isRefreshing = it
                }
                .addToDisposable(disposables)
        }

        this.viewModel.outputs.activity()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { discoveryActivitySampleAdapter.takeActivity(it) }
            .addToDisposable(disposables)

        this.viewModel.outputs.clearActivities()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { discoveryActivitySampleAdapter.takeActivity(null) }
            .addToDisposable(disposables)

        this.viewModel.outputs.startHeartAnimation()
            .compose(Transformers.observeForUIV2())
            .filter { !(lazyHeartCrossFadeAnimation()?.isRunning?:false) }
            .subscribe { lazyHeartCrossFadeAnimation()?.start() }
            .addToDisposable(disposables)

        this.viewModel.outputs.projectList()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { discoveryProjectCardAdapter.takeProjects(it) }
            .addToDisposable(disposables)

        this.viewModel.outputs.shouldShowEditorial()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { discoveryEditorialAdapter.setShouldShowEditorial(it) }
            .addToDisposable(disposables)

        this.viewModel.outputs.shouldShowEmptySavedView()
            .compose(Transformers.observeForUIV2())
            .subscribe {
                binding?.discoveryEmptyView?.isGone = !it
            }
            .addToDisposable(disposables)

        this.viewModel.outputs.shouldShowOnboardingView()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { discoveryOnboardingAdapter.setShouldShowOnboardingView(it) }
            .addToDisposable(disposables)

        this.viewModel.outputs.showActivityFeed()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startActivityFeedActivity() }
            .addToDisposable(disposables)

        this.viewModel.outputs.startSetPasswordActivity()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startSetPasswordActivity(it) }
            .addToDisposable(disposables)

        this.viewModel.outputs.startEditorialActivity()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startEditorialActivity(it) }
            .addToDisposable(disposables)

        this.viewModel.outputs.startUpdateActivity()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startUpdateActivity(it) }
            .addToDisposable(disposables)

        this.viewModel.outputs.startProjectActivity()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startProjectActivity(it.first, it.second) }
            .addToDisposable(disposables)

        this.viewModel.outputs.startPreLaunchProjectActivity()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startPreLaunchProjectActivity(it.first, it.second) }
            .addToDisposable(disposables)

        this.viewModel.outputs.showLoginTout()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startLoginToutActivity() }
            .addToDisposable(disposables)

        binding?.discoveryHeartsContainer?.let {
            it.setOnClickListener {
                this.viewModel.inputs.heartContainerClicked()
            }
        }

        this.viewModel.outputs.startLoginToutActivityToSaveProject()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { this.startLoginToutActivity() }
            .addToDisposable(disposables)

        this.viewModel.outputs.scrollToSavedProjectPosition()
            .filter { it != -1 }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                binding?.discoveryRecyclerView?.smoothScrollToPosition(it)
            }
            .addToDisposable(disposables)

        this.viewModel.outputs.showSavedPrompt()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { this.showStarToast() }
            .addToDisposable(disposables)
    }

    override fun onPause() {
        super.onPause()
        heartsAnimation = null
    }

    override fun onDetach() {
        super.onDetach()
        binding?.discoveryRecyclerView?.adapter = null
        recyclerViewPaginator?.stop()
    }

    val isAttached: Boolean
        get() = viewModel != null

    val isInstantiated: Boolean
        get() = binding?.discoveryRecyclerView != null

    private val editorialImageView: ImageView?
        get() {
            val layoutManager = binding?.discoveryRecyclerView?.layoutManager as? LinearLayoutManager
            if (layoutManager != null && discoveryEditorialAdapter != null) {
                for (i in layoutManager.findFirstVisibleItemPosition()..layoutManager.findLastVisibleItemPosition()) {
                    val childView = layoutManager.getChildAt(i)
                    if (childView != null) {
                        val viewHolder = binding?.discoveryRecyclerView?.getChildViewHolder(childView)
                        if (viewHolder is EditorialViewHolder) {
                            return childView.findViewById(R.id.editorial_graphic)
                        }
                    }
                }
            }
            return null
        }

    private fun lazyHeartCrossFadeAnimation(): AnimatorSet? {
        if (heartsAnimation == null) {
            binding?.discoveryEmptyHeartOutline?.let { discoveryEmptyHeartOutline ->
                binding?.discoveryEmptyHeartFilled?.let {
                    heartsAnimation = crossFadeAndReverse(discoveryEmptyHeartOutline, it, 400L)
                }
            }
        }
        return heartsAnimation
    }

    private fun startActivityFeedActivity() {
        context?.let {
            startActivity(Intent(context, ActivityFeedActivity::class.java))
        }
    }

    private fun startSetPasswordActivity(email: String) {
        context?.let {
            val intent = Intent().getSetPasswordActivity(requireContext(), email)
            startActivityForResult(intent, ActivityRequestCodes.LOGIN_FLOW)
            TransitionUtils.transition(requireContext(), TransitionUtils.fadeIn())
        }
    }

    private fun showStarToast() {
        context?.let {
            ViewUtils.showToastFromTop(
                requireContext(),
                getString(this.projectStarConfirmationString),
                0,
                resources.getDimensionPixelSize(R.dimen.grid_8)
            )
        }
    }

    private fun startEditorialActivity(editorial: Editorial) {
        val activity = activity
        // The transition view must be an ImageView
        val editorialImageView = editorialImageView
        if (activity != null && editorialImageView != null) {
            val intent = Intent(activity, EditorialActivity::class.java)
                .putExtra(IntentKey.EDITORIAL, editorial)
            val options = ActivityOptions.makeSceneTransitionAnimation(activity, editorialImageView, "editorial")
            startActivity(intent, options.toBundle())
        }
    }

    private fun startLoginToutActivity() {
        context?.let {
            val intent = Intent(context, LoginToutActivity::class.java)
                .putExtra(IntentKey.LOGIN_REASON, LoginReason.DEFAULT)
            startActivityForResult(intent, ActivityRequestCodes.LOGIN_FLOW)
            TransitionUtils.transition(it, TransitionUtils.fadeIn())
        }
    }

    private fun startPreLaunchProjectActivity(project: Project, refTag: RefTag) {
        context?.let {
            val intent = Intent().getPreLaunchProjectActivity(requireContext(), project.slug())
                .putExtra(IntentKey.REF_TAG, refTag)
                .putExtra(
                    IntentKey.PREVIOUS_SCREEN,
                    ThirdPartyEventValues.ScreenName.DISCOVERY.value
                )
            startActivity(intent)
            TransitionUtils.transition(requireContext(), TransitionUtils.slideInFromRight())
        }
    }

    private fun startProjectActivity(project: Project, refTag: RefTag) {
        context?.let {
            val intent = Intent().getProjectIntent(it)
                .putExtra(IntentKey.PROJECT_PARAM, project.slug())
                .putExtra(IntentKey.REF_TAG, refTag)
                .putExtra(IntentKey.PREVIOUS_SCREEN, ThirdPartyEventValues.ScreenName.DISCOVERY.value)
            startActivity(intent)
            TransitionUtils.transition(it, TransitionUtils.slideInFromRight())
        }
    }

    private fun startUpdateActivity(activity: Activity) {
        context?.let {
            val intent = Intent(context, UpdateActivity::class.java)
                .putExtra(IntentKey.PROJECT, activity.project())
                .putExtra(IntentKey.UPDATE, activity.update())
            startActivity(intent)
            TransitionUtils.transition(it, TransitionUtils.slideInFromRight())
        }
    }

    fun refresh() {
        this.viewModel.inputs.refresh()
    }

    fun takeCategories(categories: List<Category>) {
        this.viewModel.inputs.rootCategories(categories)
    }

    fun updateParams(params: DiscoveryParams) {
        this.viewModel.inputs.paramsFromActivity(params)
    }

    fun clearPage() {
        this.viewModel.inputs.clearPage()
    }

    fun scrollToTop() {
        binding?.discoveryRecyclerView?.smoothScrollToPosition(0)
    }

    companion object {
        @JvmStatic
        fun newInstance(position: Int): DiscoveryFragment {
            val fragment = DiscoveryFragment()
            val bundle = Bundle()
            bundle.putInt(ArgumentsKey.DISCOVERY_SORT_POSITION, position)
            fragment.arguments = bundle
            return fragment
        }
    }
}
